google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -437,6 +449,13 @@ public TransactionContext begin() {
         try {
           return internalBegin();
         } catch (SessionNotFoundException e) {
+          sessionReplacementCount++;
+          if (sessionReplacementCount > MAX_SESSION_REPLACES) {
+            throw SpannerExceptionFactory.newSpannerException(
+                e.getErrorCode(),
+                "Could not replace invalidated session because the maximum number of times the session may be replaced was exceeded",
```

Comment:
This is not a good experience for our customers since this degrades their read / writes. 

When would this happen ?

Reply:
Sorry, this shouldn't have been in this PR and has been removed.

It could happen if a large number of sessions in the pool have been invalidated by the server without the client knowing about it, and this would safeguard against a large number of retries that all fail.

Reply:
sessions in the pool are kept alive using "GetSession() rpc" or "SELECT 1" query.

Hopefully we shouldnt have this catastrophic event